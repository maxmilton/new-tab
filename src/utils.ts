// NOTE: When updating also update references that get an item by index
export const DEFAULT_SECTION_ORDER = [
  'Open Tabs',
  'Bookmarks',
  'History',
  'Top Sites',
  'Recently Closed Tabs',
];

export const handleClick = (event: MouseEvent): void => {
  const { target, ctrlKey } = event;
  const url = (target as HTMLAnchorElement).href;

  // only apply special handling to non-http links
  if (url && !url.startsWith('h')) {
    event.preventDefault();

    if ((target as HTMLAnchorElement).target === '_blank' || ctrlKey) {
      // open the location in a new tab
      void chrome.tabs.create({ url });
    } else {
      // update the location in the current tab
      void chrome.tabs.update({ url });
    }
  }

  // focus search element on bad click
  if(target.tagName=== 'BODY') {
    document.getElementById('s').focus();
  }
};
