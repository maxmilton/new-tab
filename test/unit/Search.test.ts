import { afterEach, beforeEach, expect, test } from 'bun:test';
import { cleanup, render } from './utils';

// HACK: The Search component is designed to be rendered once (does not clone
// its view), for byte savings. Given its mutation of the view (affecting global
// state) when run, it's vital to reset its module state between tests to
// maintain accurate test conditions.
const MODULE_PATH = import.meta.resolveSync('../../src/components/Search');
let Search: typeof import('../../src/components/Search').Search;

beforeEach(async () => {
  Loader.registry.delete(MODULE_PATH);
  Search = (await import('../../src/components/Search')).Search;
});

afterEach(cleanup);

// XXX: Because async chrome APIs are called within Search, we need to wait for
// them to complete before we assert anything.

test('rendered DOM contains expected elements', async () => {
  const rendered = render(Search());
  await happyDOM.whenAsyncComplete();
  const root = rendered.container.firstChild as HTMLElement;
  expect(root).toBeInstanceOf(window.HTMLDivElement);
  expect(root.id).toBe('c');
  const input = root.querySelector('input#s');
  expect(input).toBeTruthy();
  expect(root.firstChild).toBe(input);
  expect(input?.parentElement).toBe(root);
  const icon = root.querySelector('svg#i');
  expect(icon).toBeTruthy();
  expect(icon?.parentElement).toBe(root);

  // TODO: Check for other elements (but probably only those which are part of
  // the Search component and not its children?).
});

test('rendered DOM matches snapshot', async () => {
  const rendered = render(Search());
  await happyDOM.whenAsyncComplete();
  expect(rendered.container.innerHTML).toMatchSnapshot();
});

// TODO: Test with various settings, especially section order
// TODO: Test the search functionality
