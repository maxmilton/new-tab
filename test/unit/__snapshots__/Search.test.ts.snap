// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`rendered DOM matches snapshot 1`] = `"<div id="c"><input id="s" type="search" placeholder="Search browser..."><svg id="i"><circle cx="11" cy="11" r="8"></circle><line x1="24" y1="24" x2="16.65" y2="16.65"></line></svg><div hidden=""><h2>Open Tabs</h2><div></div><button>Show more</button></div><div hidden=""><h2>Bookmarks</h2><div></div><button>Show more</button></div><div hidden=""><h2>History</h2><div></div><button>Show more</button></div><div hidden=""><h2>Top Sites</h2><div></div><button>Show more</button></div><div hidden=""><h2>Recently Closed Tabs</h2><div></div><button>Show more</button></div></div>"`;
